src/dune_file_watcher/dune_file_watcher.ml

```diff
@@ -169,14 +175,27 @@ module For_tests = struct
   let should_exclude = should_exclude
 end
 
-let process_inotify_event (event : Async_inotify_for_dune.Async_inotify.Event.t)
-    : Event.t list =
+let process_inotify_event ~inotify
+    (event : Async_inotify_for_dune.Async_inotify.Event.t) :
+    Event.t list * Path.t list =
   let create_event_unless_excluded ~kind ~path =
     match should_exclude path with
-    | true -> []
+    | true -> ([], [])
     | false ->
       let path = Path.of_string path in
-      [ Event.Fs_memo_event (Fs_memo_event.create ~kind ~path) ]
+      let inotify = Fdecl.get inotify in
+      Mutex.lock inotify.mutex;
+      let to_scan =
+        match Table.mem inotify.awaiting_creation path with
+        | false -> None
+        | true ->
+          Inotify_lib.add inotify.inotify (Path.to_string path);
```

Comment:
The assumption here seems to be that the `path` got created. Maybe worth adding an `assert` about this?

Reply:
Actually, it doesn't seem great that we're doing all this for every event, not just for events that correspond to path creations. I think it would be nice to refactor this logic so that `inotify.awaiting_creation` and the associated mutex is only touched when necessary.